### PR TITLE
now-cli: init at 11.4.6

### PR DESCRIPTION
**Pull request by brendan-hall**
###### Motivation for this change
Adds Zeit's now-cli, per #36395 (sorry @giodamelio)
Uses the binary distribution, which presented some challenges as it's essentially an archive of a nodejs, and all the js files, with hardcoded offsets to said files.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] ~~Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))~~
- [ ] ~~Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`~~
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by brendan-hall** [comment]
I was quite surprised to find that genericBuild can't handle plain .gz files.
Have I written the unpack correctly? I couldn't find anything stating how to leave the downloaded files or the newly unpacked.

**Comment by grahamc** [comment]
@GrahamcOfBorg build now-cli

File: pkgs/development/web/now-cli/default.nix [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+    # now is a node program packaged using zeit/pkg.
+    # thus, it contains hardcoded offsets.
+    # patchelf shifts these locations when it expands headers.
+
+    # this could probably be generalised into allowing any program packaged
+    # with zeit/pkg to be run on nixos.
+
+  preFixup = let
+    libPath = lib.makeLibraryPath [stdenv.cc.cc.lib];
```
Comment (Infinisil) [inline pkgs/development/web/now-cli/default.nix]: `makeLibraryPath` already takes the `.lib` output if it exists, so you can remove that.

File: pkgs/development/web/now-cli/default.nix [hunk 2]
```diff
@@ -0,0 +1,68 @@
+{ stdenv, lib, makeWrapper, fetchurl }:
```
Comment (Infinisil) [inline pkgs/development/web/now-cli/default.nix]: `makeWrapper` is unused

**Review by infinisil — CHANGES_REQUESTED** [review]

**Comment by brendan-hall** [comment]
@Infinisil RE Your review: How's that?

File: pkgs/development/web/now-cli/default.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    new_size=$(stat --printf=%s $out/bin/now)
+
+    ###### zeit-pkg fixing starts here.
+    # we're replacing plaintext js code that looks like
+    # PAYLOAD_POSITION = '1234                  ' | 0
+    # [...]
+    # PRELUDE_POSITION = '1234                  ' | 0
+    # ^-----20-chars-----^^------22-chars------^
+    # ^-- grep points here
```
Comment (Infinisil) [inline pkgs/development/web/now-cli/default.nix]: Nice!

**Review by infinisil — APPROVED** [review]
One thing left to do for the future is to add macOS support too. This is looking good as it is now. when you squash the last commit I'll merge this.

**Comment by Hinidu** [comment]
Guys this is awesome - you've just saved my weekend!
I had exactly the same problem with packaging new version of Unity https://github.com/NixOS/nixpkgs/issues/34399
I've just copied your `zeit/pkg` fix and it finally works now!
It can't be a coincidence :smile:

**Comment by brendan-hall** [comment]
Unfortunately, I don't have a Mac. How much of the existing work can be shared with a macOS system? (I know that Zeit provide a macOS binary explicitly)

**Comment by Hinidu** [comment]
@brendan-hall they use `zeit/pkg` for Unity Package Manager since version 2017.4 I suppose. And the editor can't launch without it.

File: pkgs/development/web/now-cli/default.nix (hunk 3 above)
Comment (dch) [reply pkgs/development/web/now-cli/default.nix:50]: it's been a while but here is an alternative solution, maybe its useful

```perl
perl -pe 's/(var PAYLOAD_POSITION = .)(\d+)/$1 . ($2 + 4096)/e; s/(var PRELUDE_POSITION = .)(\\d+)/$1 . ($2 + 4096)/e;' < tailwind-freebsd-x64 > tailwind-damien
```

This relies on rpath injection causing a 4096 byte extension obviously.

works on FreeBSD.

**Pull request merged**